lib/pavilion/lockfile.py

```diff
@@ -328,3 +329,99 @@ def __exit__(self, exc_type, exc_val, exc_tb):
 
         self._done_event.set()
         self._thread.join()
+
+
+class FuzzyLock:
+    """A custom lock object designed for use on NFS systems. The lock behaves like a queue:
+    each process declares its intent to take the lock, and the lock goes to whichever process
+    declared first. The object lessens, but does not fully resolve, some of the concurrency issues
+    related to NFS, particularly in the case where multiple Pavilion instances are working with
+    the same build. Intended to be invoked as a context manager, using the 'with' keyword.
+    """
+    def __init__(self, lock_dir: Path, wait_time: float = 0.5,
+     timeout: Optional[float] = None):
+        """
+        :param lock_dir: directory in which lockfiles will be created
+        :param wait_time: time to wait between checking status
+        :param timeout: time (in seconds) after which the lock times out.
+            None may be passed to indicate no timeout.
+        """
+        self._lock_dir = lock_dir
+        self._wait_time = wait_time
+        self._timeout = timeout
+
+        self._mtimes = {} # type: Dict[Path, float]
+
+        # create a (statistically) unique lockfile
+        self._lockfile = lock_dir / f"{uuid4().hex[-16:]}.lock"
+
+    def __enter__(self) -> 'FuzzyLock':
+        """
+        Attempt to acquire the lock.
+
+        :return: True if lock has been successfully acquired; False if timed out
+        """
+        self._lock_dir.mkdir(exist_ok=True)
+
+        # Declare intent to take lock
+        self._lockfile.touch(exist_ok=False)
+
+        first = False
+        start = time.time()
+
+        while not first:
+            # Give other instances opportunity to declare intent
+            sleep(self._wait_time)
+
+            if self._timeout is not None and self._timeout >= 0 and \
+                (time.time() - start > self._timeout):
+                raise TimeoutError("Timeout while attempting to acquire lock")
+
+            earliest = self._get_earliest()
+            first = earliest == self._lockfile
+
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        # Remove the lockfile, since no longer competing for lock
+
+        self._lockfile.unlink()
+        del self._mtimes[self._lockfile]
+
+        # Remove lock directory once it's empty
+        if len(list(self._lock_dir.iterdir())) == 0:
```

Comment:
Catch a generic OSError.  No need to check if directory is empty before trying to delete it.